### ISSUE TITLE
~20% slower than defaults build

### ISSUE THREAD

**Issue by dougalsutherland** [conda-forge/python-feedstock#169]
Using python 3.6 on Linux, `pyperformance` gives significantly slower speeds on a variety of benchmarks with `conda-forge`'s package than with the one from `defaults`:

![histogram; values range from ~ 1.1 to 1.4](https://imgur.com/x7IxsZ4.png)

This might just be from the new compilers, but some of the new optimizations may only be possible because [`defaults` statically links the `python` executable](https://groups.google.com/a/continuum.io/forum/#!topic/anaconda/057P4uNWyCU).

Maybe we should be doing that, and/or also enabling more compiler optimizations. (Probably should wait to work on that until conda-forge is using the updated compilers, though.)

**Comment by dougalsutherland** [comment]
(This came up from https://github.com/shogun-toolbox/shogun/issues/4068, when a user used the conda-forge shogun package with a `defaults` build of python, and got a segfault because of the static linking issue. That was really a shogun bug, but making this change might also make the two more compatible.)

**Comment by msarahan** [comment]
Yes, the easiest path forward is to get conda-forge up to date using conda-build 3 and the new compilers, and then merge Defaults' python recipe over.

@mingwandroid spent a lot of time optimizing things.  Part of it is the static linkage, but there's also some PGO/LTO stuff going on.

**Comment by mingwandroid** [comment]
Also good compilers and -O3 (do not use that for general builds though).

**Comment by mingwandroid** [comment]
BTW the ~20% speed difference is the same on macOS (for Python 3.*) and Windows (for Python 3.6).

**Comment by marscher** [comment]
on these platforms recent compilers are used, so they can not be the origin of this issue here?!

**Comment by mingwandroid** [comment]
Conda forge do not use good compilers but anyway the various comments in this thread detail the other parts of the speedup.

**Comment by mingwandroid** [comment]
@mrschar, to be specific this thread has already mentioned:

1. LTO
2. PGO
3. Building a static interpreter

So I am not sure why you focus only on the compilers used (which in conda-forge's case are both poor and old).

**Comment by jakirkham** [comment]
Is this still true of more recent builds? Specifically ones with the new compilers?

**Comment by dougalsutherland** [comment]
I just ran with the latest builds of 3.7.1 and actually got that conda-forge has gotten *much* worse compared to defaults. Possible that I'm doing something wrong or it was a fluke on my machine; I was doing normal usage at the same time. I ran the cf tests twice in case something stupid happened there, but it was pretty consistent; only ran defaults once.

Also, I'm installing `perf` and `performance` with pip, because they're in `defaults` but not in conda-forge but I only checked conda-forge. :stuck_out_tongue:

```bash
$ conda create -n py-defaults -c defaults --override-channels python
$ conda activate py-defaults
$ pip install performance  # is in defaults but not conda-forge, so just install from pip for consistency
$ python -m performance run -o py-defaults.json  # wait 45 minutes...

$ conda create -n py-cf -c conda-forge --override-channels python
$ conda activate py-cf
$ pip install performance
$ python -m performance run -o py-cf.json  # wait even longer...

$ conda activate py3  # back to env with plotting and stuff
$ pip install perf  # also in defaults but not conda-forge
$ python -m perf compare_to py-defaults.json py-cf.json --table | grep -o '[-\+].*\%' | tr -d '+%' | python -c 'import sys; import numpy as np; d = np.array([int(l.strip()) / 100. + 1 for l in sys.stdin if l.strip()]); import matplotlib as mpl; mpl.use("Agg"); import matplotlib.pyplot as plt; import seaborn as sns; sns.set(); sns.distplot(d); plt.xlabel("conda-forge slowdown"); plt.savefig("comparison.png");'
```

![comparison](https://user-images.githubusercontent.com/36478/49169536-da869400-f331-11e8-8945-79b5c3a71887.png)

**Comment by dougalsutherland** [comment]
Unless I'm doing something wrong @jjhelmus or just got fabulously unlucky, it seems like this is actually a bigger problem now than it was before.

**Comment by dougalsutherland** [comment]
Oh, oops, of course.

Will report back in 45 minutes....

**Comment by mingwandroid** [comment]
> Unless I'm doing something wrong @jjhelmus or just got fabulously unlucky, it seems like this is actually a bigger problem now than it was before, so I'm reopening.

Does this mean defaults Python has gotten even faster? Excellent news if so.

**Comment by dougalsutherland** [comment]
Much better! We seem the same as `defaults` on the `gcc7` builds.

![comparison](https://user-images.githubusercontent.com/36478/49174686-cc8b4000-f33e-11e8-8530-8642ff292e8f.png)

@mingwandroid I haven't rerun old `defaults` out of laziness, but I think that's probably true.

**Comment by mingwandroid** [comment]
Great news!
